# amr_project
![image](https://github.com/hophuquy/amr_project/blob/main/Soil_pipeline.png)
Prerequisites
Before executing these commands, ensure the following prerequisites are met:
  * Ç¹extflow is installed on your system.
  * conda is installed on your system.
  * Python(version: Python 3.10.13)  is available.
  * R(version: R version 4.3.3) is available.
  * The necessary data files for the mag and funcscan workflow are accessible.
  * before run this pipeline you have to run requirements.py to install third-party software for this analysis.

# Full analysis in Colab
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1YGVOXaKbDMg_U8yGJJ4mxY1r-z9LttZq#scrollTo=qogN1-XmJSwr) Notebook

## Upstream Analysis

### Steps:
1. Filter and convert result of funscan to reference
  * create folder for analysis
  * hamronization filter
2. Mapping
  * create gff file
  * mapping using bowties2
  * retrieve only mapped read
3. Create and Normalize feature table using using tpm normalization.
  * create feature count table
  * tpm normalization
  * get amr drug tpm abundace table

### Tools:
- Python(version: Python 3.10.13)
- R(version: R version 4.3.3)
- Bioinformatics Libraries: bowtie2, featureCounts

## Downstream Analysis

### Steps:
1. Alpha beta diversity analysis
2. Visualize abundace using bar plot
3. Detect duplicate  amr gennes in all sample identity 75%, 90%
  * detect pathogen and their amr genes
  * Get amr ontology using card database
  * get amr drug tpm abundace table
4. Build phylogeny tree of pathogen detected
5. Abundance different analysis

### Tools:
- Python(version: Python 3.10.13)
- R(version: R version 4.3.3)
- Bioinformatics Libraries
